### ISSUE TITLE
Add Distribute API to check for updates

### ISSUE THREAD

**Issue by andrekiba** [microsoft/appcenter-sdk-dotnet#758]
### **Description**

Following the guide on site I set up in-app updates for a Xamarin.Forms (only Android for now) app.
I register the callback just before the `AppCenter.Start` inside `App.xaml.cs OnStart()` method.
I precisly need to know when the callback is supposed to be called: every time the app starts? every time the activity resume? ... The documentation is not clear at all on this point :-)
This is because the app is designed to work in single application mode inside a kiosk and never reboots, nor is the activity changed. In HockeyApp you could decide when to check for new updates, here how I can do it?
I know I'm forced to interact with the user to be able to update the app but I want this is the only case.
In this scenario it seems also that `MandatoryUpdate` doesn't work because it is always false, also when I distributed a mandatory build (in our case all distributed builds will be mandatory).
Finally...please provide a way to te test in-app updates in debug otherwise it is not comfortable at all... 

Thank you! :-)

### **Details**

1. What is your app platform (Xamarin.Android or Xamarin.iOS or UWP)?
    Xamarin.Forms with Android only target
2. If using Xamarin.Forms or if using portable/shared code to call our SDK APIs, are you using shared project, PCL code or .NET standard code for the application? Which .NET standard version or which PCL profile?
    .NET standard 2.0.
2. Which SDK version are you using?
    1.9.0
4. Which OS version did you experience the issue on?
    the device has Android 7.1.0, project compiled with 8.1 Api 27 SDK
5. What device version did you see this error on?  Were you using an emulator or a physical device?
    Physical device Lenovo Tab 4 10

**Comment by achocron** [comment]
As of now, updates are checked when the Distribute service starts. However, while we don't have an API to check for available updates, you can work around this by disabling and then re-enabling the Distribute service.

```csharp
            await Distribute.SetEnabledAsync(false);
            await Distribute.SetEnabledAsync(true);
```

**Comment by andrekiba** [comment]
Thank you @achocron for now I try with the workaround. For the future it will be nice to don’t check at all. Something like: call the callback when new release is available for the distribution group and the app (the device) is in that group. I think you have an “event” when the user from the portal (or from the rest api) creates a new release.

**Comment by andrekiba** [comment]
For mandatory updates yes, exactly what you said, it seems that they are always not mandatory.

**Comment by andrekiba** [comment]
Another point is that the automatic update procedure seems very slow: it takes more than a minute from when the user chooses to update (by sending “UpdateAction.Update” notification to the AppCenter) and when the popup to start up the update procedure is shown. In hockeyapp was practically immediate and at the end the application restarted automatically...here no, another prompt to the user.

**Comment by clpolet** [comment]
The download of the new version happens when the “UpdateAction.Update” is notified to the SDK. Then it depends on the network, weight of the application...

@andrekiba what do you exactly mean by the mandatory updates are not mandatory? In the release details passed to the `onReleaseAvailable` do you see the `MandatoryUpdate` property being true?

**Comment by clpolet** [comment]
@andrekiba sorry for the late response. About the first point (mandatory update bug) you should ask the AppCenter support request (blue button bottom right of the AC portal). It might not be just an SDK issue.
Let me double check about the no user feedback on update that seems odd.

**Comment by clpolet** [comment]
I did some testing on our Xamarin Form test application to verify the behavior, as you said the SDK is taking care of the download after the `UpdateAction.Update` but it does trigger a system notification in the notification tray that shows the progress of the download. When the download is done then Android would popup the update dialog (without me doing anything) to update the application.

So you are saying that you don't see the system notification that shows the progress of the download?

Can you please share a code snippet of your customization so we can try to reproduce the issue. Also when you disable your customization are you seeing the progress of the download?

Thanks.

**Comment by andrekiba** [comment]
thank you for your reply @clpolet.
Yes the behavior that you have described it really happens but I cannot understand what are the benefits compared to the old why what we have in Hockeyapp (progress dialog in app). Myabe in my application I don't want to use notifications, maybe the application is in single task mode (dedicated device) where notifications are disabled by default.
Is there a way to understand when the downlaod is done from within the app?
For the point related to mandatory updates I try to  rise an AppCenter support request.

**Comment by clpolet** [comment]
@andrekiba I don't think it's doable from what we expose today. Let me tag this as a feature flag.

**Comment by guperrot** [comment]
Today the mandatory updates download is shown as a modal progress dialog to prevent the user from using the app until the update is installed.

If the update is optional (default) the progress is not getting in the way, since we use the system download manager of Android the notification is actually managed by the system and you can still use the app during download.

Hockey App had issues with storage permissions and cleaning the files, that was fixed by switching to the system download manager.

**Comment by andrekiba** [comment]
thank you very much for your explanation @guperrot. In this case the behavior is perfect and yes we need to understand why the release is not mandatory. I can only say that I set the mandatory flag on the portal but when the callback is called the release is always not mandatory :-(

**Comment by ElektrojungeMS** [comment]
I apologize for the inconvenience but can you please get in touch through Intercom/Support (blue button on the portal) so we can share more info in private and look at this end to end from both an SDK and a backend perspective? Please mention this GH issue and that you've been talking to the SDK crew.
Thanks in advance!

**Comment by guperrot** [comment]
I tested on one of our testing app and the mandatory feature still works as I explained.
To be sure, please call `AppCenter.LogLevel = LogLevel.Verbose` before AppCenter.Start then in logcat logs you should see the following logs when the check is made:

```

**Comment by andrekiba** [comment]
sorry @guperrot In which way I can see this log? updates are disabled in debug mode...and the log works only in debug mode... what I missed?

**Comment by guperrot** [comment]
Logs can be visible on a release if you call `AppCenter.LogLevel = LogLevel.Verbose` before `AppCenter.start`.

Then you use the **Device Monitor** in **Tools** (for VS for Mac, the windows menu might be slightly different) to look at the device logs and filter by `AppCenter` regular expression in log cat.

**Comment by guperrot** [comment]
Can you also confirm that you created a support ticket on http://appcenter.ms feedback button so that the server team can help on this?

**Comment by andrekiba** [comment]
@guperrot using your suggestion I can clearly see the logcat and, as supposed, the release is not mandatory
`10-23 17:23:49.465 | AmbrogioLog | OnReleaseAvailable id=25 version=35 mandatory=False`

**Comment by guperrot** [comment]
The API to check for updates is now available in version [3.1.0](https://github.com/microsoft/appcenter-sdk-dotnet/releases/tag/3.1.0).
